core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -752,6 +768,14 @@ class GroupCoordinator(val brokerId: Int,
 
   private def removeMemberAndUpdateGroup(group: GroupMetadata, member: MemberMetadata, reason: String) {
     group.remove(member.memberId)
+
+    if (member.awaitingJoinCallback != null) {
+      // This can happen if a new member times out before the group can finish the rebalance.
+      // We return UNKNOWN_MEMBER_ID so that the consumer will retry the JoinGroup request if
+      // is still active.
+      member.awaitingJoinCallback(joinError(NoMemberId, Errors.UNKNOWN_MEMBER_ID))
```

Comment:
Should we just call `invokeJoinCallback` which does other things like setting the callback to null, decrementing `numMembersAwaitingJoin` as well?

Reply:
The reason I invoked the callback directly is that the member has already been removed from the group. I think that's probably why I didn't both setting the callback to null as well. I think we can just change the callback first. Then all paths go through `invokeJoinCallback`.